src/data/real/ereal.lean

```diff
@@ -0,0 +1,159 @@
+
+/-- - -a = a on ereal -/
+theorem ereal.neg_neg : ∀ (a : ereal), - (- a) = a
+| none := rfl
+| (some none) := rfl
+| (some (some a)) := show (((- -a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)) = (((a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)),
+by simp [neg_neg a]
+
+/-- a ≤ -b → b ≤ -a on ereal -/
+theorem ereal.le_neg_of_le_neg {a b : ereal} (h : a ≤ -b) : b ≤ -a :=
+by rwa [←ereal.neg_neg b, ereal.neg_le, ereal.neg_neg]
+
+/-- The claim that a set of ereals has a supremum in ereal -/
+def has_Sup (X : set ereal) : Prop := ∃ l : ereal, is_lub X l
+
+local attribute [instance, priority 10] classical.prop_decidable
+
+/-- A set of ereals has a Sup in ereal -/
+theorem Sup_exists (X : set ereal) : has_Sup X :=
+  let Xoc : set (with_top ℝ) := λ x, X (↑x : with_bot _) in
+dite (Xoc = ∅) (λ h, ⟨⊥, ⟨
+    by
+    { rintro (⟨⟩|x) hx, exact le_refl none,
+      exfalso,
+      apply set.not_mem_empty x,
+      rw ←h,
+      exact hx,
+    },
+    λ u hu, lattice.bot_le⟩
+  ⟩) (λ h, dite (⊤ ∈ Xoc) (λ h2, ⟨⊤, ⟨λ _ _, lattice.le_top, λ x hx, hx _ h2⟩⟩) $
+    λ htop, 
+    let Xoo : set ℝ := λ (x : ℝ), Xoc (↑ x) in
+    begin
+    by_cases h2 : nonempty (upper_bounds Xoo),
+    { rcases h2 with ⟨b, hb⟩,
+      use (↑(↑(real.Sup Xoo : real) : with_top ℝ) : with_bot (with_top ℝ)),
+      split,
+      { rintros (⟨⟩|⟨⟩|x) hx,
+            exact lattice.bot_le,
+          exact false.elim (htop hx),
+        change (↑(↑x : with_top ℝ) : with_bot (with_top ℝ)) ≤ _,
+        simp [real.le_Sup _ ⟨b, hb⟩ hx],
+      },
+      { intros c hc,
+        cases c with c,
+          cases (set.exists_mem_of_ne_empty h) with x hx,
+          cases (lattice.le_bot_iff.1 (hc (↑x : with_bot _) hx)),
+        cases c with c, {unfold_coes, simp},
+        suffices : real.Sup Xoo ≤ c,
+          unfold_coes, simp [this],
+        refine (real.Sup_le Xoo _ ⟨b, hb⟩).2 _,
+          rcases (set.exists_mem_of_ne_empty h) with ⟨⟨⟩ | ⟨x⟩, hx⟩,
+            contradiction,
+          exact ⟨x, hx⟩,
+        intros x hx,
+        replace hc := hc (↑(↑x : with_top ℝ) : with_bot (with_top ℝ)) hx,
+        unfold_coes at hc,
+        simpa using hc,
+      }
+    },
+    { use ⊤,
+      split, intros x hx, exact lattice.le_top,
+      intros b hb,
+      rw lattice.top_le_iff,
+      cases b with b,
+        exfalso,
+        apply h,
+        ext x,
+        split, swap, rintro ⟨⟩,
+        intro hx,
+        cases (lattice.le_bot_iff.1 (hb (↑x : with_bot _) hx)),
+      cases b with b, refl,
+      exfalso,
+      apply h2,
+      use b,
+      intros x hx,
+      replace hb := hb (↑(↑x : with_top ℝ) : with_bot (with_top ℝ)) hx,
+      unfold_coes at hb,
+      simpa using hb,
+    }
+  end) 
+
+noncomputable def ereal.Sup := λ X, classical.some (Sup_exists X)
+
+noncomputable instance : lattice.has_Sup ereal := ⟨ereal.Sup⟩
+
+/-- `ereal` is a complete lattice -/
+noncomputable instance : lattice.complete_lattice (ereal) :=
```

Comment:
This instance might as well be generalised to `with_top (with_bot ` of any `conditionally_complete_lattice`

Reply:
I think it would be nicer to write with a with_top_and_bot type with three constructors and <= defined by cases and without this \in trickery (which bites you twice in the current approach).

Reply:
I agree that it would be better to first do this for a general `conditionally_complete_lattice` and then instantiate it to `real`.